### PR TITLE
Update libcontainer to include PRs with fixes to systemd cgroup driver

### LINKED ISSUE #61474: ./hack/local-up-cluster.sh fails on Fedora rawhide
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

I attempted to start all-on-one testing environment with

```
ENABLE_DAEMON=true PATH='$(pwd)/kubernetes/third_party/etcd:${PATH}' ./hack/local-up-cluster.sh
```

on Fedora rawhide, the way it works just fine on Fedora 27. The node never becomes `Ready`.

**What you expected to happen**:

The `./hack/local-up-cluster.sh` finishing and the node `127.0.0.1` becoming `Ready` shortly after.

**How to reproduce it (as minimally and precisely as possible)**:

```
git clone https://github.com/kubernetes/kubernetes.git
( cd kubernetes && ./hack/install-etcd.sh )
( cd kubernetes && ENABLE_DAEMON=true PATH='$(pwd)/kubernetes/third_party/etcd:${PATH}' ./hack/local-up-cluster.sh )
kubectl get nodes
```

**Anything else we need to know?**:

On Fedora 27, these steps give me working setup.

The `kubectl describe node 127.0.0.1` says (among others)
```
Conditions:
  Type             Status    LastHeartbeatTime                 LastTransitionTime                Reason                    Message
  ----             ------    -----------------                 ------------------                ------                    -------
  OutOfDisk        Unknown   Wed, 21 Mar 2018 09:23:21 -0400   Wed, 21 Mar 2018 09:24:04 -0400   NodeStatusUnknown         Kubelet stopped posting node status.
  MemoryPressure   Unknown   Wed, 21 Mar 2018 09:23:21 -0400   Wed, 21 Mar 2018 09:24:04 -0400   NodeStatusUnknown         Kubelet stopped posting node status.
  DiskPressure     Unknown   Wed, 21 Mar 2018 09:23:21 -0400   Wed, 21 Mar 2018 09:24:04 -0400   NodeStatusUnknown         Kubelet stopped posting node status.
  PIDPressure      False     Wed, 21 Mar 2018 09:23:21 -0400   Wed, 21 Mar 2018 09:23:21 -0400   KubeletHasSufficientPID   kubelet has sufficient PID available
  Ready            Unknown   Wed, 21 Mar 2018 09:23:21 -0400   Wed, 21 Mar 2018 09:24:04 -0400   NodeStatusUnknown         Kubelet stopped posting node status.
```

The `/tmp/kubelet.log` ends with
```
I0321 09:23:23.492042   27717 factory.go:110] Factory "systemd" can handle container "/system.slice/dev-hugepages.mount", but ignoring.
I0321 09:23:23.492059   27717 factory.go:110] Factory "systemd" can handle container "/system.slice/tmp.mount", but ignoring.
I0321 09:23:23.496752   27717 cpu_manager.go:155] [cpumanager] starting with none policy
I0321 09:23:23.496810   27717 cpu_manager.go:156] [cpumanager] reconciling every 10s
I0321 09:23:23.496830   27717 policy_none.go:42] [cpumanager] none policy: Start
I0321 09:23:23.517797   27717 factory.go:113] Using factory "raw" for container "/libcontainer_27717_systemd_test_default.slice"
I0321 09:23:23.518259   27717 manager.go:997] Added container: "/libcontainer_27717_systemd_test_default.slice" (aliases: [], namespace: "")
I0321 09:23:23.518614   27717 container.go:448] Start housekeeping for container "/libcontainer_27717_systemd_test_default.slice"
I0321 09:23:23.521188   27717 manager.go:1054] Destroyed container: "/libcontainer_27717_systemd_test_default.slice" (aliases: [], namespace: "")
E0321 09:23:23.527391   27717 node_container_manager.go:52] Failed to create "/kubepods" cgroup
F0321 09:23:23.527441   27717 kubelet.go:1363] Failed to start ContainerManager Delegation not available for unit type
```

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0-alpha.0.1341+c5d4a032d73f90", GitCommit:"c5d4a032d73f90f286fde1a7ca3e782538c47f58", GitTreeState:"clean", BuildDate:"2018-03-21T13:21:57Z", GoVersion:"go1.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0-alpha.0.1341+c5d4a032d73f90", GitCommit:"c5d4a032d73f90f286fde1a7ca3e782538c47f58", GitTreeState:"clean", BuildDate:"2018-03-21T13:21:57Z", GoVersion:"go1.10", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
```
I0321 09:23:21.140790   27717 manager.go:227] Machine: {NumCores:8 CpuFrequency:2412325 MemoryCapacity:16817651712 HugePages:[{PageSize:2048 NumPages:0}] MachineID:f72669acdc2840be9a3ca8a2019e3fdb SystemUUID:00020003-0004-0005-0006-000700080009 BootID:39dff605-2d5a-4abc-9cc8-c26c44c0c601 Filesystems:[{Device:tmpfs DeviceMajor:0 DeviceMinor:20 Capacity:8408825856 Type:vfs Inodes:2052936 HasInodes:true} {Device:/dev/mapper/fedora-root DeviceMajor:253 DeviceMinor:0 Capacity:16095641600 Type:vfs Inodes:7864320 HasInodes:true} {Device:/dev/sda1 DeviceMajor:8 DeviceMinor:1 Capacity:1023303680 Type:vfs Inodes:65536 HasInodes:true}] DiskMap:map[253:1:{Name:dm-1 Major:253 Minor:1 Size:8476688384 Scheduler:none} 253:2:{Name:dm-2 Major:253 Minor:2 Size:897581056 Scheduler:none} 253:3:{Name:dm-3 Major:253 Minor:3 Size:346738917376 Scheduler:none} 253:4:{Name:dm-4 Major:253 Minor:4 Size:346738917376 Scheduler:none} 8:0:{Name:sda Major:8 Minor:0 Size:896998047744 Scheduler:cfq} 253:0:{Name:dm-0 Major:253 Minor:0 Size:16106127360 Scheduler:none}] NetworkDevices:[{Name:enp2s0 MacAddress:00:d0:68:10:03:66 Speed:1000 Mtu:1500} {Name:enp3s0 MacAddress:00:d0:68:10:03:67 Speed:-1 Mtu:1500}] Topology:[{Id:0 Memory:4159594496 Cores:[{Id:0 Threads:[0] Caches:[{Size:65536 Type:Data Level:1} {Size:65536 Type:Instruction Level:1} {Size:1048576 Type:Unified Level:2}]} {Id:1 Threads:[1] Caches:[{Size:65536 Type:Data Level:1} {Size:65536 Type:Instruction Level:1} {Size:1048576 Type:Unified Level:2}]}] Caches:[]} {Id:1 Memory:4227317760 Cores:[{Id:0 Threads:[2] Caches:[{Size:65536 Type:Data Level:1} {Size:65536 Type:Instruction Level:1} {Size:1048576 Type:Unified Level:2}]} {Id:1 Threads:[3] Caches:[{Size:65536 Type:Data Level:1} {Size:65536 Type:Instruction Level:1} {Size:1048576 Type:Unified Level:2}]}] Caches:[]} {Id:2 Memory:4203909120 Cores:[{Id:0 Threads:[4] Caches:[{Size:65536 Type:Data Level:1} {Size:65536 Type:Instruction Level:1} {Size:1048576 Type:Unified Level:2}]} {Id:1 Threads:[5] Caches:[{Size:65536 Type:Data Level:1} {Size:65536 Type:Instruction Level:1} {Size:1048576 Type:Unified Level:2}]}] Caches:[]} {Id:3 Memory:4226830336 Cores:[{Id:0 Threads:[6] Caches:[{Size:65536 Type:Data Level:1} {Size:65536 Type:Instruction Level:1} {Size:1048576 Type:Unified Level:2}]} {Id:1 Threads:[7] Caches:[{Size:65536 Type:Data Level:1} {Size:65536 Type:Instruction Level:1} {Size:1048576 Type:Unified Level:2}]}] Caches:[]}] CloudProvider:Unknown InstanceType:Unknown InstanceID:None}
```
- OS (e.g. from /etc/os-release):
```
NAME=Fedora
VERSION="29 (Rawhide)"
ID=fedora
VERSION_ID=29
PRETTY_NAME="Fedora 29 (Rawhide)"
ANSI_COLOR="0;34"
CPE_NAME="cpe:/o:fedoraproject:fedora:29"
HOME_URL="https://fedoraproject.org/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=rawhide
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=rawhide
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
```
- Kernel (e.g. `uname -a`):
```
Linux redacted.example.com 4.16.0-0.rc5.git3.1.fc29.x86_64 #1 SMP Fri Mar 16 15:28:11 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools:
- Others:

```
systemd-238-4.fc29.x86_64
```

### PR DESCRIPTION
**Pull request by filbranden**
**What this PR does / why we need it**:

PR opencontainers/runc#1754 works around an issue in manager.Apply(-1) that makes Kubelet startup hang when using systemd cgroup driver (by adding a timeout) and further PR opencontainers/runc#1772 fixes that bug by checking the proper error status before waiting on the channel.
    
PR opencontainers/runc#1776 checks whether Delegate works in slices, which keeps libcontainer systemd cgroup driver working on systemd v237+.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61474

**Special notes for your reviewer**:
/assign @derekwaynecarr
cc @vikaschoudhary16 @sjenning @adelton @mrunalp 

**Release note**:

```release-note
Fixes a hang on kubelet startup when using systemd cgroup driver of libcontainer.
```

### CODE REVIEW

**Comment by filbranden** [comment]
/assign @sttts 
for OWNERS

**Comment by filbranden** [comment]
I'll probably want to include opencontainers/runc#1772 as well which has a better fix for this...

Maybe we can hold this PR until that one is pushed upstream too.

**Comment by filbranden** [comment]
@derekwaynecarr Please take another look, this includes all the fixes we pushed to libcontainer. (You might want to retest #61474 yet one more time, but I imagine everything should be fine.)

We need approval from dep-approvers, so I'll assign the three of them here...

/assign @cblecker 
/assign @thockin
/assign @sttts 

Cheers!
Filipe

**Comment by cblecker** [comment]
I can approve the dependency bump, but it should be reviewed by sig node first (@derekwaynecarr) and tests should pass.

**Comment by filbranden** [comment]
Tests are all good... @derekwaynecarr you're next :smiley:

**Comment by dims** [comment]
LGTM 👍

**Comment by mrunalp** [comment]
:+1:

**Comment by mrunalp** [comment]
@runcom @umohnani @dustymabe fyi

**Comment by umohnani8** [comment]
@mrunalp works for me!

File: vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go
Comment (derekwaynecarr) [inline vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go]: @filbranden we hit an issue with this in production and @sjenning @vikaschoudhary16 need to push a revert.

"In https://github.com/opencontainers/runc/pull/1683 and follow on https://github.com/opencontainers/runc/pull/1754, we attempted to handle a race condition https://github.com/openshift/origin/issues/16246 by waiting for systemd to ack the creation of the transient unit created by StartTransientUnit().

This did solve the race but unfortunately started the code down a path in the systemd dbus library that I'm now convinced was never tested.

If you pass a channel to StartTransientUnit() it takes the jobListener lock in startJob() where it didn't before.  This causes the deadlock situation when even one systemd ack is missed (for yet unknown reasons)."

see: https://github.com/openshift/origin/pull/19043

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
i would like @sjenning and @vikaschoudhary16 to comment on the this PR.  we hit an issue with this change recently in production on rhel/centos versions of systemd.  i do not think they hit the upstream systemd, but i honestly cannot recall.  i would like them to comment, but we may need to revert the one libcontainer pr.

File: vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go
Comment (sjenning) [reply vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go]: @filbranden yes, this has been a mess.  We ended up reverting both  opencontainers/runc#1683  and  opencontainers/runc#1754 in the vendored runc for openshift/origin v3.9 because of the deadlock we could encounter in the `coreos/go-systemd/dbus` package. See `startJob()` and `jobComplete()` w/ the `jobListener` lock.

https://bugzilla.redhat.com/show_bug.cgi?id=1558183#c5

Unfortunately, it looks like we dropped to ball getting that revert upstream.   The channel wait still exists in runc master, kube master and 1.10, and openshift/origin master as we just rebased on kube 1.10 and resynced our deps.

The whole "passing a channel to StartTransientUnit()" need to be rolled back everywhere.

Comment (filbranden) [reply vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go]: Ok sounds good, let's hold on this.

Before we revert everything... May I suggest we spend some time trying to understand whether there's a bug in `coreos/go-systemd/dbus` that we can fix instead?

I'll cc myself to the Bugzilla in RedHat.

Thanks!
Filipe

**Comment by vikaschoudhary16** [comment]
@derekwaynecarr @sjenning 
>  i do not think they hit the upstream systemd

 right, this issue was with rhel version systemd only and not with upstream systemd. And now it has got fixed in rhel version systemd as well, therefore there is no side-effect of reverting this change. I will soon create PRs to revert it from everywhere.
Thanks!

**Comment by filbranden** [comment]
Ok, so I opened issue opencontainers/runc#1780 to track this. I think that is probably more appropriate than here.

I tried to summarize the history of events that brought us here.

Regarding the solution, I disagree that we should revert all the PRs. I believe that making StartTransientUnit synchronous is the right call, it's the more appropriate behavior, and we should keep that.

I believe there's one remaining bug, when we have a timeout in StartTransientUnit, the code writing to the channel will block forever. There's an easy fix/workaround for that, by making the channel buffered, with a buffer of one element, that way the write won't block. I'll write a PR for that.

I think there's some further cleanup we might want to do there, but I'd start there.

Cheers,
Filipe

**Comment by cblecker** [comment]
/uncc brendandburns jbeda
/unassign cblecker sttts thockin 

Will leave this up to sig-node to determine a direction. Please assign back if this becomes ready for approval.

**Comment by filbranden** [comment]
@vikaschoudhary16 @sjenning @mrunalp Can you please review opencontainers/runc#1781 ?

If that looks good, then I'd like to include it here and then finally merge this into Kubernetes code base...

Thanks!
Filipe

File: vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go
Comment (filbranden) [reply vendor/github.com/opencontainers/runc/libcontainer/cgroups/systemd/apply_systemd.go]: This is fixed in opencontainers/runc#1781.

**Comment by filbranden** [comment]
Ok I updated this PR with all the fixes.

Considering the current state is "broken" (as far as I can tell, Kubernetes upstream doesn't work at all on a system using systemd cgroup driver), I think we should really go ahead and submit this.

If you want to take a short while to try to reproduce the issue (particularly the one when the systemd response to dbus times out) please go ahead, but let's try to figure this one quickly and merge it before new conflicts arise please 😄 

Cheers,
Filipe

**Review by derekwaynecarr — APPROVED** [review]

**Comment by filbranden** [comment]
/assign @cblecker 
for approval...

**Comment by cblecker** [comment]
Existing dep bump, with no license changes.
/approve

**Comment by cblecker** [comment]
/hold
There's an issue with deps on master.

**Pull request merged**
